### PR TITLE
feat(managed-data): support for more functionality

### PR DESCRIPTION
**Pull request by sg-s**
## changes

- [x] support for a bunch of new endpoints: creating/updating/deleting workspaces/databases
- [x] use interrogate in github actions to check for docstrings
- [x] docstrings for ALL functions
- [x] tests for above 
- [x] expanded schema to support responses from this
- [x] CLI support this new functionality
- [x] overhauled CLI to use human-friendly verbs

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -50,6 +50,15 @@ jobs:
           version: 0.3.1
           args: check --select I
 
+      - name: Check docstrings using interrogate
```
Comment (sg-s) [inline .github/workflows/main.yml:53]: we're using interrogate to force ourselves to write docstrings for every function and module

File: scripts/build_docs.sh [hunk 2]
```diff
@@ -20,11 +20,12 @@ if [ "$CI" = "true" ]; then
 else
   echo "Running Locally, will not install."
   source ./venv/bin/activate && \
-      MKDOCS_OUT="$(mkdocs build -s 2>&1)" \
+      MKDOCS_OUT="$(mkdocs build -s 2>&1)" && \
```
Comment (sg-s) [inline scripts/build_docs.sh:23]: this variable wasn't correctly being set

File: src/cli/data.py [hunk 3]
```diff
@@ -89,155 +91,158 @@ def merge_db(self):
...
         arguments=[
             (
-                ["file_id"],
-                {"help": "File ID", "action": "store"},
+                ["source"],
+                {"help": "Source", "action": "store"},
             ),
-        ],
-    )
-    def download_file(self):
```
Comment (sg-s) [inline src/cli/data.py:100]: deleted because subsumed in the copy command

File: src/cli/data.py [hunk 4]
```diff
@@ -89,155 +91,158 @@ def merge_db(self):
...
+                ["--files"],
                 {
                     "action": "store_true",
-                    "help": "Whether to return JSON formatted data [default: False]",
+                    "help": "Whether to list files: [False]",
                 },
             ),
-        ],
-    )
-    def describe_row(self):
```
Comment (sg-s) [inline src/cli/data.py:150]: deleted because this is handled in the generic describe command

File: src/cli/data.py [hunk 5]
```diff
@@ -89,155 +91,158 @@ def merge_db(self):
...
+        if self.app.pargs.files:
+            # we will only list files
+            data = _api.list_files(
+                client=self._get_client(),
+            )
+            if not self.app.pargs.json:
+                # show a table with file names, ids, status
+                pdata = dict(Name=[], Status=[], ID=[])
+                for item in data:
+                    pdata["Name"].append(item["file"]["name"])
```
Comment (sg-s) [inline src/cli/data.py:184]: i'm choosing to only show some things in the "pretty" table

File: src/cli/data.py [hunk 6]
```diff
@@ -248,85 +253,44 @@ def list_files(self):
...
-        """list rows"""
+    def describe(self):
+        """describe file or row"""
 
-        rows = _api.list_rows(
-            parent_id=self.app.pargs.row_id,
-            client=self._get_client(),
-        )
+        if PREFIXES.FILE in self.app.pargs.object_id:
+            data = _api.describe_file(
```
Comment (sg-s) [inline src/cli/data.py:260]: assuming that if a system ID begins with "_file", then it's a file

File: src/cli/data.py [hunk 7]
```diff
@@ -337,72 +301,30 @@ def ls(self):
...
+        data = _api.describe_row(
+            row_id=self.app.pargs.object_id,
             client=self._get_client(),
         )
+        row_type = data["type"]
 
-        _print_dict(row_data, json=self.app.pargs.json)
-
-
-class CopyController(cement.Controller):
```
Comment (sg-s) [inline src/cli/data.py:350]: not sure why this was a controller, every subcommand is now a function

File: src/managed_data/_api.py [hunk 8]
```diff
@@ -32,6 +36,243 @@ def _get_default_client(client: Optional[Client] = None):
...
+        hid: Human ID of the workspace.
+        name: Name of the workspace.
+        parent_id: ID (or human ID) of the parent.
+
+    Returns:
+        A dictionary that conforms to a [CreateWorkspaceResponse][src.managed_data.schema.CreateWorkspaceResponse].
+    """
+    client = _get_default_client(client)
+
+    if hid is None:
```
Comment (sg-s) [inline src/managed_data/_api.py:61]: this allows you to simply use this with a name

File: src/managed_data/_api.py [hunk 9]
```diff
@@ -32,6 +36,243 @@ def _get_default_client(client: Optional[Client] = None):
...
+        parent_id: ID of the parent workspace.
+        hid_prefix: Human ID prefix of the database. This prefix is used in every row.
+
+
+    Returns:
+        A dictionary that conforms to a [CreateDatabaseResponse][src.managed_data.schema.CreateDatabaseResponse]."""
+
+    client = _get_default_client(client)
+
+    if hid is None:
```
Comment (sg-s) [inline src/managed_data/_api.py:99]: allows you to skip hid

File: src/managed_data/_api.py [hunk 10]
```diff
@@ -32,6 +36,243 @@ def _get_default_client(client: Optional[Client] = None):
...
+
+    return client.invoke("UpdateDatabase", data)
+
+
+@beartype
+def delete_rows(
+    row_ids: list[str],
+    *,
+    client: Optional[Client] = None,
+) -> None:
```
Comment (sg-s) [inline src/managed_data/_api.py:188]: nothing is returned because the API returns nothing

File: src/managed_data/api.py [hunk 11]
```diff
@@ -7,17 +7,10 @@
 
 from beartype import beartype
 from deeporigin.exceptions import DeepOriginException
-from deeporigin.managed_data._api import (
-    convert_id_format,
```
Comment (sg-s) [inline src/managed_data/api.py:11]: i don't want to pollute the namespace so importing _api

File: src/managed_data/client.py [hunk 12]
```diff
@@ -19,18 +22,40 @@
...
+        id="_file:placeholder-file",
+        uri="s3://placeholder/uri",
+        name="placeholder",
+        status="ready",
+        contentLength=123,
+        contentType="application/foo",
+        dateCreated="2024-05-08 19:16:09.26078",
+        dateUpdated="2024-05-08 19:16:09.26078",
+        createdByUserDrn="drn:identity::user:google-apps|foo@deeporigin.com",
+    ).model_dump()
```
Comment (sg-s) [inline src/managed_data/client.py:38]: switching to model_dump because dict() is deprecated

File: src/managed_data/client.py [hunk 13]
```diff
@@ -176,10 +210,27 @@ def invoke_list_rows(self, data: dict):
                     id=self.databases[0],
                     type="database",
                     parentId=self.workspaces[0],
-                ).dict()
+                ).model_dump()
+            ]
+        elif data == dict(filters=[dict(rowType="row")]):
+            # return the example row
```
Comment (sg-s) [inline src/managed_data/client.py:216]: this was missing previously. only affects tests

File: src/managed_data/client.py [hunk 14]
```diff
@@ -315,14 +396,3 @@ def _check_response(response: requests.models.Response) -> Union[dict, list]:
         return response["data"]
     else:
         raise KeyError("`data` not in response")
-
-
-def file_description():
-    return dict(
```
Comment (sg-s) [inline src/managed_data/client.py:321]: moved this, now uses the basemodel

File: src/managed_data/schema.py [hunk 15]
```diff
@@ -8,7 +8,7 @@
 
 from typing import Literal, Optional, Union
 
-from pydantic import BaseModel
+from pydantic import BaseModel, ConfigDict
```
Comment (sg-s) [inline src/managed_data/schema.py:11]: using this instead of class-based config

File: src/utils.py [hunk 16]
```diff
@@ -12,7 +13,15 @@
 ]
 
 
-PREFIX = "deeporigin://"
+@dataclass
+class PREFIXES:
```
Comment (sg-s) [inline src/utils.py:17]: defined all constants prefixes here. this allows me to refer to any prefix easily using `PREFIXES.DO` e.g.

File: tests/test_cli_data.py [hunk 17]
```diff
@@ -10,15 +11,25 @@
     Client,
     DeepOriginClient,
     MockClient,
-    file_description,
 )
+from deeporigin.managed_data.schema import DescribeFileResponse, ListRowsResponse
 
-describe_file_inputs = [
```
Comment (sg-s) [inline tests/test_cli_data.py:16]: this wasn't being used

File: tests/test_cli_data.py [hunk 18]
```diff
@@ -32,9 +43,11 @@ def config(pytestconfig):
     if pytestconfig.getoption("client") == "mock":
         data["client"] = MockClient()
 
-        data["databases"] = ["db-sample"]
-        data["rows"] = ["row-1"]
-        data["file"] = file_description()
+        # unpack mock data from client
```
Comment (sg-s) [inline tests/test_cli_data.py:46]: source of truth now is the client

File: tests/test_cli_data.py [hunk 19]
```diff
@@ -67,78 +80,65 @@ def test_data(config):
     assert "List data in managed data on Deep Origin" in stdout, "Unexpected output"
 
 
-def test_describe_file(config):
+@pytest.mark.parametrize("json_option", JSON_OPTIONS)
```
Comment (sg-s) [inline tests/test_cli_data.py:83]: note this allows us to test every CLI command with both the normal (human) output, and `--json` 

this happens automatically -- pytest handles this for us

File: tests/test_cli_data.py [hunk 20]
```diff
@@ -67,78 +80,65 @@ def test_data(config):
...
+    if option == ["--json"]:
+        data = _check_json(stdout)
 
+        assert data["id"] == row_id, "Expected row ID to match"
 
-def test_list_rows(config):
-    db_id = config["databases"][0]
 
+@pytest.mark.parametrize("json_option", JSON_OPTIONS)
+@pytest.mark.parametrize("list_option", LIST_OPTIONS)
```
Comment (sg-s) [inline tests/test_cli_data.py:115]: the list command has a bunch of options, and we use a single function to test all of them

File: tests/test_managed_data.py [hunk 21]
```diff
@@ -60,6 +65,38 @@ def config(pytestconfig):
     # teardown tasks, if any
 
 
+def test_create_workspace(config):
+    data = _api.create_workspace(
+        name="test-" + str(uuid.uuid4())[:6],
+        client=config["client"],
+    )
+
+    CreateWorkspaceResponse(**data)
```
Comment (sg-s) [inline tests/test_managed_data.py:74]: we check the structure of the response by coercing it into that model. in future these APIs will return typed info, so we won't have to do this

**Review by sg-s — COMMENTED** [review]
self-review

File: src/managed_data/schema.py [hunk 22]
```diff
@@ -35,13 +36,51 @@
...
+Cardinality = Literal["one", "many"]
+
 IDFormat = Literal["human-id", "system-id"]
 """Format of an ID"""
 
 DatabaseReturnType = Literal["dataframe", "dict"]
 """Return type of a database"""
 
 
+class CreateDatabaseResponse(BaseModel):
```
Comment (kennyjwilli) [inline src/managed_data/schema.py:49]: Do you think this kind of stuff could be generated from an openapi schema in the future?

**Review by kennyjwilli — APPROVED** [review]
Awesome!

**Pull request merged**